For #11358 - Show crash notification for BACKGROUND_CHILD process crashes.

jamienicol opened the pull request:
This replaces the isFatal field of NativeCodeCrash with a processType
field which is read from the GeckoView crash intent. This can have one
of the following values:

* MAIN: indicating a fatal crash in the main process.
* FOREGROUND_CHILD: indicating a crash in a foreground child process (such as a
  Gecko content process). The application may be able to recover from this.
* BACKGROUND_CHILD: indicating a crash in a background child process (such as
  Gecko's GPU process). The engine will automatically recover from this, and the
  crash will have barely been noticeable to the user, if at all.

We treat MAIN crashes as we did fatal crashes before: we do not send the intent
to the application, but do show the prompt or notification to the user, if
required.

We treat FOREGROUND_CHILD crashes as we did non-fatal crashes before: sending
the intent to the application if requested.

For BACKGROUND_CHILD crashes, we do _not_ send the intent to the application. As
the engine automatically recovers there is no need for the application to handle
the crash. Additionally, we choose to display the crash notification rather than
prompt, as we do not wish to interrupt the user's flow.



### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

rocketsroger commented on components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt:
```diff
@@ -159,7 +159,7 @@ class MozillaSocorroService(
             extrasFilePath = crash.extrasPath,
             isNativeCodeCrash = true,
-            isFatalCrash = crash.isFatal,
+            isFatalCrash = crash.processType == Crash.NativeCodeCrash.PROCESS_TYPE_MAIN,
```
Shouldn't we pass `processType` in and send it through `sendCrashData`?

jamienicol replied:
Do we need any Socorro changes to send a new property?

Should we stop sending `isFatal` and send `processType` instead? Or do both?

rocketsroger replied:
Yes, we'll need Socorro side change after this to show the new property.  I think we should send both, this way we can still filter for fatal/non-fatal crashes and also (after the server side change) see what type of crash it is.

rocketsroger commented on components/lib/crash/src/main/java/mozilla/components/lib/crash/CrashReporter.kt:
```diff
@@ -266,8 +266,10 @@ class CrashReporter(
+            // If this is a native code crash in a foreground child process then we can recover
+            // and can notify the app. Background child process crashes will be recovered from
+            // automatically, and main process crashes cannot be recovered from.
+            crash is Crash.NativeCodeCrash && crash.processType == Crash.NativeCodeCrash.PROCESS_TYPE_FOREGROUND_CHILD
```
Do you mean `crash.processType != Crash.NativeCodeCrash.PROCESS_TYPE_MAIN`.  

There's a lot of place we're checking if it's fatal.  Is is another good reason to keep the `isFatal` property so we don't have to update these checks.

jamienicol replied:
In this case I actually do mean `crash.processType == Crash.NativeCodeCrash.PROCESS_TYPE_FOREGROUND_CHILD`.

Agi suggested that we use the crash notification for background child crashes, which as I understand it is handled in android-components rather than by the application via the intent. So we only want to send intent for FOREGROUND_CHILD crashes.

That does seem fairly arbitrary, though. Perhaps it would be better if we send the intent for all non-fatal crashes (both FOREGROUND_CHILD and BACKGROUND_CHILD). And if we expose a public API for showng the crash notification, then Fenix's intent handler could call that API in the BACKGROUND_CHILD case?

That would seem more consistent to me, what do you think?

rocketsroger replied:
Ah, I see.  It does make sense to only notify the app side when it's a FOREGROUND_CHILD crash.  I'm OK with how you've implemented it.  Thanks

rocketsroger reviewed (commented):
Looks good.  I have some questions.  Is there a easy way for me to trigger `BACKGROUND_CHILD` crashes?

jamienicol commented on components/lib/crash/src/main/java/mozilla/components/lib/crash/notification/CrashNotification.kt:
```diff
@@ -47,8 +47,16 @@ internal class CrashNotification(
 
         val channel = ensureChannelExists(context)
 
+        val title = when ((crash as? Crash.NativeCodeCrash)?.processType) {
```
You can tell I'm not a kotlin expert :smile: will make the change

jamienicol commented on components/lib/crash/src/main/java/mozilla/components/lib/crash/Crash.kt:
```diff
@@ -78,17 +79,16 @@ sealed class Crash {
         val minidumpSuccess: Boolean,
         val extrasPath: String?,
-        val isFatal: Boolean,
+        @ProcessType val processType: String?,
```
Okay. Does it sound okay to have it as a function, calculated from `processType == MAIN`? (Can it even be a getter property so syntactically there's no change? Does the fact this is a data class prevent that?)

The `fatal` argument from the geckoview crash intent is deprecated and instead we use the process type. I agree `isFatal` is useful within android-components code, but it'd be nice to have it calculated from the process type to avoid any source of inconsistency.

rocketsroger replied:
Yes, that's a good idea.  Setting `isFatal` with a function is a good solution.  This way, we don't have to break the API and also we can simplify a lot of checks to see if the crash is fatal.  Thanks

jamienicol commented:
Thanks for the review @rocketsroger! Have replied inline.

csadilek reviewed (commented):
Thank you for putting up the patch. I tested the different scenarios incl. GPU crashes and it's all working well!

I agree to keep `isFatal` which will also prevent breaking API, and we can land this tomorrow without further coordination (no Fenix side changes needed).

A more general question is if this non-intrusive way to notify users will be good enough i.e., will we see enough submissions just based on the system notifications, or do we need something more. This is big improvement though overall, so thanks again!

Mugurell commented:
Also tested on Focus and saw this working nicely 👍

jamienicol commented:
Latest push addresses all the review comments. Except I chose _not_ to submit the `process type` field to Socorro, as we already send the process type to Socorro anyway (I believe in the "extras" file). Let me know if that sounds okay

rocketsroger reviewed (approved):
LGTM! 🚢  Thanks!
